collector/src/execute/profiler.rs

```diff
@@ -226,7 +226,46 @@ impl<'a> Processor for ProfileProcessor<'a> {
                 let cgout_file = filepath(self.output_dir, &out_file("cgout"));
                 let cgann_file = filepath(self.output_dir, &out_file("cgann"));
 
-                fs::copy(&tmp_cgout_file, &cgout_file)?;
+                // It's useful to filter all `file:function` entries from
+                // jemalloc into a single fake
+                // `<all-jemalloc-files>:<all-jemalloc-functions>` entry. That
+                // way the cost of all allocations is visible in one line,
+                // rather than spread across many small entries.
+                //
+                // The downside is that we don't get any annotations within
+                // jemalloc source files, but this is no real loss, given that
+                // jemalloc is basically a black box whose code we never look
+                // at anyway. DHAT is the best way to profile allocations.
+                let file = fs::File::open(&tmp_cgout_file)?;
+                let reader = io::BufReader::new(file);
+                let mut cgout = fs::File::create(&cgout_file).unwrap();
```

Comment:
Maybe a `BufWriter` should also be used here? The files can probably be big (?) and we write to them line by line.

Reply:
Good suggestions! I have also added a line to `ci/check-profiling.sh` to look for the `<all-jemalloc-files>:<all-jemalloc-functions>` line, as a basic test it's working.

Reply:
BTW, this got me thinking about buffered/unbuffered readers and writers, and made me realise I hadn't fully understood some of their details. I wrote a new paragraph in the [I/O chapter of the perf-book](https://nnethercote.github.io/perf-book/io.html) as a result:

> Forgetting to buffer is more common when writing. Both unbuffered and buffered writers implement the [Write](https://doc.rust-lang.org/std/io/trait.Write.html) trait, which means the code for writing to an unbuffered writer and a buffered writer is much the same. In contrast, unbuffered readers implement the [Read](https://doc.rust-lang.org/std/io/trait.Read.html) trait but buffered readers implement the [BufRead](https://doc.rust-lang.org/std/io/trait.BufRead.html) trait, which means the code for reading from an unbuffered reader and a buffered reader is different. For example, it is difficult to read a file line by line with an unbuffered reader, but it is trivial with a buffered reader by using [BufRead::read_line](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.read_line) or [BufRead::lines](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines). For this reason, it is hard to write an example for readers like the one above for writers, where the before and after versions are so similar.